### ISSUE TITLE
NetlibIncomplete is not caught when reading SOCKS handshake

### ISSUE THREAD

**Issue by taiyangc** [mitmproxy/mitmproxy#593]
Encountered this while running latest mitmproxy 0.12.0 with this command:

```mitmproxy -v -e --socks --ssl-version-client all --ssl-port 14943 --tcp :14943```

Client connects to mitmproxy via socks5 using websocket (messages dumped with --tcp):

```
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/libmproxy/proxy/server.py", line 73, in handle
    upstream_info = self.config.mode.get_upstream_server(self.client_conn)
  File "/Library/Python/2.7/site-packages/libmproxy/proxy/primitives.py", line 95, in get_upstream_server
    client_greet = socks.ClientGreeting.from_file(client_conn.rfile)
  File "/Library/Python/2.7/site-packages/netlib/socks.py", line 64, in from_file
    ver, nmethods = struct.unpack("!BB", f.safe_read(2))
  File "/Library/Python/2.7/site-packages/netlib/tcp.py", line 207, in safe_read
    "Expected %s bytes, got %s"%(length, len(result))
NetLibIncomplete: Expected 2 bytes, got 0

mitmproxy has crashed!
Please lodge a bug report at: https://github.com/mitmproxy/mitmproxy
```

The event log:

```
192.168.101.57:45408: Establish SSL
  -> with client
  -> with server (sni: None)
192.168.101.57:45408: SSL handshake error: The client may not trust the proxy's certificate.
192.168.101.57:45408: ProxyError('NetLibError("SSL handshake error: SysCallError(-1, \'Unexpected EOF\')",)',)
192.168.101.57:45408: serverdisconnect
  -> [IP]:14943
192.168.101.57:45409: clientconnect
192.168.101.57:45409: Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/libmproxy/proxy/server.py", line 73, in handle
    upstream_info = self.config.mode.get_upstream_server(self.client_conn)
  File "/Library/Python/2.7/site-packages/libmproxy/proxy/primitives.py", line 95, in get_upstream_server
    client_greet = socks.ClientGreeting.from_file(client_conn.rfile)
  File "/Library/Python/2.7/site-packages/netlib/socks.py", line 64, in from_file
    ver, nmethods = struct.unpack("!BB", f.safe_read(2))
  File "/Library/Python/2.7/site-packages/netlib/tcp.py", line 207, in safe_read
    "Expected %s bytes, got %s"%(length, len(result))
NetLibIncomplete: Expected 2 bytes, got 0
```

I'm not sure what this means but a crash doesn't seem right. I can reproduce this so if there is anything I should provide, let me know.

---

I have a second question related to #427: I get the same SSL handshake error. The server I connect to does have HSTS policy. But I thought HSTS is client enforced. After I disable all hostname checks (Java), I still get the same error. I must be overlooking something... maybe someone can enlighten me.

**Comment by mhils** [comment]
Hi there,

the second client in your log (192.168.101.57:45409) doesn't send any data and immediately disconnects instead. This has unrelated to the previous SSL connection, which comes from a different TCP stream (:4540**8**). This is nothing we can fix, but I'll leave this issue open to remind me to improve the error message.

Cheers,
Max

**Comment by cortesi** [comment]
Just committed a partial fix for this in mitmproxy/netlib@f7bd690e3aba0be05c30a3b9a4d499de8dbd5e06. This changes the error we raise when we get 0 bytes to a Disconnect, which I believe we handle correctly. We still need to handle a genuine incomplete read too, and give a non-traceback error.
